ast/dml.go

```diff
@@ -85,7 +85,59 @@ type Join struct {
 
 // Restore implements Node interface.
 func (n *Join) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	if ctx.JoinLevel != 0 {
+		ctx.WritePlain("(")
+		defer ctx.WritePlain(")")
+	}
+	ctx.JoinLevel++
```

Comment:
could we use type asserts to decide whether to add brackets

Reply:
Actually I tried as you suggest, but it is hard to handle the following case. For example, based on current parser, the ASTs of `select * from t1, t2` and `select * from (t1) join t2` are the same:
```
*ast.SelectStmt	&{{{{select * from t1,t2}}} {[]} 0xc00000f590 false 0xc00000a720 <nil> 0xc00000f5c0 <nil> <nil> [] <nil> <nil> none [] false false}
  *ast.TableRefsClause	&{{} 0xc000148100}
    *ast.Join	&{{} {[]} 0xc000148080 0xc000052380 1 <nil> [] false false}
      *ast.Join	&{{} {[]} 0xc000052340 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000146090 }
          *ast.TableName	&{{} {[]}  t1 <nil> <nil> []}
      *ast.TableSource	&{{} 0xc000146120 }
        *ast.TableName	&{{} {[]}  t2 <nil> <nil> []}
  *ast.FieldList	&{{} [0xc000056660]}
    *ast.SelectField	&{{} 7 0xc0000a4d20 <nil>  false}
```
But the AST of `select * from (t1) join (t2)` is different:
```
...
    *ast.Join	&{{} {[]} 0xc000164080 0xc000164100 1 <nil> [] false false}
      *ast.Join	&{{} {[]} 0xc0000a2300 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000162090 }
          *ast.TableName	&{{} {[]}  t1 <nil> <nil> []}
      *ast.Join	&{{} {[]} 0xc0000a2340 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000162120 }
          *ast.TableName	&{{} {[]}  t2 <nil> <nil> []}
  ...
```
It is natural to find that `Join`s under the top `Join` should be wrapped with parentheses, and we don't have to think about other cases. If not, we have to consider the actual type of `TableSource`, which could be a normal table, a `SelectStmt` or a `UnionStmt`. I think that is more complicated.

Reply:
ok, detailed explanation, good job!